ext/limiter/limiter.c

```diff
@@ -0,0 +1,158 @@
+        return 0;
+    }
+
+    return (ts.tv_sec * NANOSECONDS_PER_SECOND) + /* seconds as nanoseconds */
+            ts.tv_nsec;                           /* plus remaining nanoseconds */
+}
+
+void ddtrace_limiter_create() {
+    uint32_t limit = (uint32_t) get_global_DD_TRACE_RATE_LIMIT();
+
+    if (!limit) {
+        return;
+    }
+
+    __limiter = mmap(NULL, sysconf(_SC_PAGESIZE), PROT_WRITE, MAP_SHARED|MAP_ANONYMOUS, -1, 0);
+
+    if ( __limiter == MAP_FAILED) {
+        __limiter = NULL;
+        return;
+    }
+
+    __limiter->limit = limit;
+    __limiter->window.open = ddtrace_limiter_clock();
+    __limiter->window.limit = limit;
+
+    pthread_mutexattr_t mattr;
+
+    pthread_mutexattr_init(&mattr);
+    pthread_mutexattr_setpshared(
+        &mattr, PTHREAD_PROCESS_SHARED);
+    pthread_mutex_init(&__limiter->lock, &mattr);
+    pthread_mutexattr_destroy(&mattr);
+}
+
+bool ddtrace_limiter_active() {
+    if (!__limiter) {
+        return false;
+    }
+
+    return true;
+}
+
+bool ddtrace_limiter_allow() {
+    ZEND_ASSERT(__limiter);
+
+    pthread_mutex_lock(&__limiter->lock);
+
+    uint64_t previous = __limiter->window.current;
+
+    __limiter->window.current  = ddtrace_limiter_clock();
+
+    uint64_t elapsed = __limiter->window.current - previous;
+
+    if (__limiter->window.current > (__limiter->window.open + NANOSECONDS_PER_SECOND)) {
+        /* window passed, reset */
+
+        /* store current window rate for effective rate calculation */
+        __limiter->window.rate = __limiter->window.samples ?
+            (double) __limiter->window.allowed / __limiter->limit : 1;
+
+        /* move window */
+        __limiter->window.open = __limiter->window.current;
+        __limiter->window.samples = 0;
+        __limiter->window.allowed = 0;
+    }
+
+    if (__limiter->window.limit != __limiter->limit) {
+        __limiter->window.limit = MIN(
+            __limiter->limit,
+            __limiter->window.limit + ((elapsed / NANOSECONDS_PER_SECOND) * __limiter->limit));
+    }
+
+    if (__limiter->window.samples++ < __limiter->window.limit) {
+        __limiter->window.allowed++;
+
+        pthread_mutex_unlock(&__limiter->lock);
+        return true;
+    }
+
+    pthread_mutex_unlock(&__limiter->lock);
+    return false;
+}
+
+double ddtrace_limiter_rate() {
+    double effective = 0;
+
+    pthread_mutex_lock(&__limiter->lock);
+
+    if (!__limiter->window.rate ||
+        !__limiter->window.samples) {
+        /* no previous window, or samples */
+        effective = (double) __limiter->window.allowed / __limiter->limit;
+
+        goto __ddtrace_limiter_rate_leave;
+    }
+
+    effective = (
+                    (double) (__limiter->window.allowed / __limiter->limit) + /* current rate */
+                    __limiter->window.rate                                    /* previous window rate */
+                ) / 2.0;                                                      /* spread over last two windows */
```

Comment:
doesn't that need to take the time elapsed into account? I.e. `1 + (window.current - window.open) / NANOSECONDS_PER_SECOND` or something like that instead of `2.0`? Otherwise there should be a steep cliff by 50% in the number (as `window.allowed` starts over at 0 again?

Reply:
this now mirrors python